plugins/workspace-plugin/src/workspace-projects-manager.ts

```diff
@@ -92,21 +92,22 @@ export class WorkspaceProjectsManager {
 
     theia.window.showInformationMessage('Che Workspace: Starting importing projects.');
 
-    const cloningPromises: PromiseLike<string>[] = [];
+    const cloningPromises: Promise<string>[] = [];
     for (const cloneCommand of cloneCommandList) {
       try {
-        const cloningPromise = cloneCommand.execute();
-        cloningPromises.push(cloningPromise);
+        let cloningPromise = cloneCommand.execute();
 
         if (isMultiRoot) {
-          cloningPromise.then(projectPath => this.workspaceFolderUpdater.addWorkspaceFolder(projectPath));
+          cloningPromise = cloningPromise.then(projectPath => {
+            this.workspaceFolderUpdater.addWorkspaceFolder(projectPath);
```

Comment:
> > If I understand correctly, the problem is: we wait for cloning operation, but it's not enough - we should wait for adding a project as workspace folder.
> > Correct?
> 
> @RomanNikitenko yes, because if we notify clients that the clone operation has been finished, they can look at current workspace folders and it will return an empty list while they expect that everything is setup.

I'm asking because for such goals `addWorkspaceFolder()` is not designed.
I mean there is no guarantee that a project has added as a workspace folder when a promise has resolved.
The promise can be resolved, for example, when [a project just added to a queue](https://github.com/eclipse-che/che-theia/blob/60773883a5ff09ec30e3853f856b72031f9f0440/plugins/workspace-plugin/src/workspace-folder-updater.ts#L26).

I have cared about logic that add projects as workspace folders step by step and with respect [the documentation](https://github.com/eclipse-theia/theia/blob/8b9fb43c763f2f1d179f9839ef993d7218d99073/packages/plugin/src/theia.d.ts#L5902) for `theia.workspace.updateWorkspaceFolders()` API.  
But I didn't care about honesty of returned promise as I didn't see the use case which you described above when I was adding `WorkspaceFolderUpdater` entity. Sorry for that...

On the other hand - maybe the current changes should fix https://github.com/eclipse/che/issues/19610 as we are waiting for all cloned projects. But looking on the `addWorkspaceFolder()` implementation - there is no guarantee...

Could you clarify how I can test the changes. Usually, I add a component with reference for the editor which contains required changes. 
I don't know how apply that approach for:

>Use a factory link like #https://github.com/spring-projects/spring-petclinic with this image

Reply:
to test the changes, you need to click on a factory link as stated in `how to test section`

`#https://github.com/spring-projects/spring-petclinic` but not on workspaces.openshift.com (need a che up-to-date instance)

Then when opening the workspaces, you'll see plug-in recommendations
if you open the same factory URL without the custom che-theia editor, you won't see plug-in recommendations

Reply:
you can look at generate PR yaml like `https://raw.githubusercontent.com/chepullreq4/pr-check-files/master/che-theia/pr-1079/simple/che-theia-simple-devfile.yaml` as well and update project to `https://github.com/spring-projects/spring-petclinic`